Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs

```diff
@@ -0,0 +1,298 @@
+﻿// <copyright file="FunctionsController.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Corvus.Testing.AzureFunctions
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.IO;
+    using System.Management;
+    using System.Threading.Tasks;
+    using Corvus.Testing.AzureFunctions.Internal;
+
+    /// <summary>
+    /// Starts, manages and terminates functions instances for testing purposes.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// This class supports a limited degree of thread safety: you can have multiple calls to
+    /// <see cref="StartFunctionsInstance(string, string, int, string, string, FunctionConfiguration?)"/> in progress simultaneously for a single
+    /// instance of this class, but <see cref="TeardownFunctions"/> must not be called concurrently
+    /// with any other calls into this class. The intention is to enable tests to spin up multiple
+    /// functions simultaneously. This is useful because function startup can be the dominant
+    /// factor in test execution time for integration tests.
+    /// </para>
+    /// </remarks>
+    public sealed class FunctionsController
+    {
+        private const long StartupTimeout = 60;
+
+        private readonly List<FunctionOutputBufferHandler> output = new List<FunctionOutputBufferHandler>();
+        private readonly object sync = new object();
+
+        /// <summary>
+        /// Start a functions instance.
+        /// </summary>
+        /// <param name="testDirectory">The directory associated with the current test context,
+        /// usually available from your testing framework's context object(s).</param>
+        /// <param name="path">The location of the functions project.</param>
+        /// <param name="port">The port on which to start the functions instance.</param>
+        /// <param name="runtime">The runtime version, defaults to netcoreapp2.1.</param>
+        /// <param name="provider">The functions provider. Defaults to csharp.</param>
+        /// <param name="configuration">A <see cref="FunctionConfiguration"/> instance, for conveying
+        /// configuration values via environment variables to the function host process.</param>
+        /// <returns>A task that completes once the function instance has started.</returns>
+        public async Task StartFunctionsInstance(
+            string testDirectory,
+            string path,
+            int port,
+            string runtime = "netcoreapp3.1",
```

Comment:
As mentioned earlier, this ended up being an optional argument because older versions of the library didn't let you specify the runtime. Since we're defining a new class, I think maybe we should make this mandatory? Thoughts @jongeorge1 ?

Reply:
Well, at the least we should make it consistent with the comment! (line 42).

The best argument I can think of for making it explicit is that it will avoid hard-to-track-down errors where it's been left to default to `netcoreapp3.1` but the target function doesn't work with that runtime for whatever reason. So going on the whole "pit of success" goal, let's remove the default.

Reply:
Done in ef8c898 (comment also updated).